Fabric.Authorization.API/scripts/AzureAD-Migration.ps1

```diff
@@ -0,0 +1,252 @@
+        do {
+            try {
+                $credential = Connect-AzureADTenant -tenantId $tenantId
+                if (!$credentials.ContainsKey($tenantId)) {
+                    Write-DosMessage -Level Information -Message "Caching credentials for tenant $($tenantId)."
+                    $credentials.Add($tenantId, $credential)
+                }
+
+                Write-DosMessage -Level "Information" -Message "Retrieving group $($groupSID) from Azure AD Tenant $($tenantId)..."
+                $azureADGroups = Get-AzureADGroup -Filter "onPremisesSecurityIdentifier eq '$($groupSID)'"
+                $authenticationFailed = $false
+            }
+            catch [Microsoft.Open.Azure.AD.CommonLibrary.AadAuthenticationFailedException], [Microsoft.Open.Azure.AD.CommonLibrary.AadNeedAuthenticationException] {
+                Write-DosMessage -Level Error -Message "Error authenticating user with Azure AD tenant $($tenantId). Please try again."
+                $authenticationFailed = $true
+            }
+            # Connect-AzureADTenant will not throw an exception if a valid username (i.e., user@domain, even if user is not in tenant
+            # and/or invalid password is entered). This catch block handles the 403 that is returned when Get-AzureADGroup fails
+            # for this scenario
+            catch [Microsoft.Open.AzureAD16.Client.ApiException] {
+                if ($_.Exception.ErrorCode -eq 403) {
+                    Write-DosMessage -Level Information -Message "403 error when retrieving Azure AD group."
+                    # reset cached credentials for this tenantId
+                    $credentials.Remove($tenantId)
+                    Write-DosMessage -Level "Information" -Message "Removed cached credentials for Azure AD Tenant $($tenantId)..."
+                    $authenticationFailed = $true
+                }
+                else {
+                    Write-DosMessage -Level Information -Message "Unexpected error when retrieving Azure AD group. $($_.Exception)"
+                    throw
+                }
+            }
+            catch {
+                $errorMsg = "Unexpected error while connecting to Azure AD: $($_.Exception)"
+                Write-DosMessage -Level Error -Message $errorMsg
+                throw $errorMsg
+            }
+        } while ($authenticationFailed -eq $true)
+    }
+
+    # disconnect from Azure AD
+    try {
+        Disconnect-AzureAD
+    }
+    catch {
+        $errorMsg = "There was an unexpected error error while disconnecting from Azure AD tenant: $($tenantId)"
+        Write-DosMessage -Level Error -Message $errorMsg
+        throw $errorMsg
+    }
+
+    if ($azureADGroups.Count -eq 1) {
+        return $azureADGroups[0]
+    }
+    elseif ($azureADGroups.Count -eq 0) {
+        $errorMsg = "No match found for SID $($groupSID) in Azure AD Tenant $($tenantId)."
+        Write-DosMessage -Level Information -Message $errorMsg
```

Comment:
Do we want this to be a fatal error? Seems like it might be better to log these and move on so we can make sure we get all the "good" ones done.